<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Promise源码</title>
  </head>

  <body>
    <!-- <script>
    // 三种状态
    const PENDING = "pending"
    const FULFILLED = "fulfilled"
    const REJECTED = "rejected"

    function MyPromise(callback) {
      var _this = this
      _this.currentState = PENDING // Promise当前的状态
      _this.value = void 0 // Promise的值
      // 用于保存 then 的回调， 只有当 promise
      // 状态为 pending 时才会缓存，并且每个实例至多缓存一个
      _this.onResolvedCallbacks = [] // Promise resolve时的回调函数集
      _this.onRejectedCallbacks = [] // Promise reject时的回调函数集
      _this.resolve = function (value) {
        if (value instanceof MyPromise) {
          // 如果 value 是个 Promise， 递归执行
          return value.then(_this.resolve, _this.reject)
        }
        setTimeout(() => { // 异步执行，保证顺序执行
          if (_this.currentState === PENDING) {
            _this.currentState = FULFILLED // 状态管理
            _this.value = value
            _this.onResolvedCallbacks.forEach(cb => cb())
          }
        })
      } // resolve 处理函数
      _this.reject = function (value) {
        setTimeout(() => { // 异步执行，保证顺序执行
          if (_this.currentState === PENDING) {
            _this.currentState = REJECTED // 状态管理
            _this.value = value
            _this.onRejectedCallbacks.forEach(cb => cb())
          }
        })
      } // reject 处理函数

      // 异常处理
      // new Promise(() => throw Error('error'))
      try {
        callback(_this.resolve, _this.reject) // 执行callback并传入相应的参数
      } catch (e) {
        _this.reject(e)
      }
    }
    // then 方法接受两个参数，onFulfilled，onRejected，分别为Promise成功或失败的回调
    MyPromise.prototype.then = function (onFulfilled, onRejected) {
      var _this = this
      // 规范 2.2.7，then 必须返回一个新的 promise
      var promise2
      // 根据规范 2.2.1 ，onFulfilled、onRejected 都是可选参数
      // onFulfilled、onRejected不是函数需要忽略，同时也实现了值穿透
      onFulfilled = typeof onFulfilled === 'function' ? onFulfilled : value => value
      onRejected = typeof onRejected === 'function' ? onRejected : error => {
        throw error
      }

      if (_this.currentState === FULFILLED) {
        // 如果promise1（此处为self/this）的状态已经确定并且为fulfilled，我们调用onFulfilled
        // 如果考虑到有可能throw，所以我们将其包在try/catch块中
        return promise2 = new MyPromise(function (resolve, reject) {
          try {
            var x = onFulfilled(_this.value)
            // 如果 onFulfilled 的返回值是一个 Promise 对象，直接取它的结果作为 promise2 的结果
            resolutionProcedure(promise2, x, resolve, reject)
          } catch (err) {
            reject(err) // 如果出错，以捕获到的错误作为promise2的结果
          }
        })
      }
      // 此处实现与FULFILLED相似，区别在使用的是onRejected而不是onFulfilled
      if (_this.currentState === REJECTED) {
        return promise2 = new MyPromise(function (resolve, reject) {
          try {
            var x = onRejected(_this.value)
            resolutionProcedure(promise2, x, resolve, reject)
          } catch (err) {
            reject(err)
          }
        })
      }
      if (_this.currentState === PENDING) {
        // 如果当前的Promise还处于PENDING状态，我们并不能确定调用onFulfilled还是onRejected
        // 只有等待Promise的状态确定后，再做处理
        // 所以我们需要把我们的两种情况的处理逻辑做成callback放入promise1（此处即_this/this）的回调数组内
        // 处理逻辑和以上相似
        return promise2 = new MyPromise(function (resolve, reject) {
          _this.onResolvedCallbacks.push(function () {
            try {
              var x = onFulfilled(_this.value)
              resolutionProcedure(promise2, x, resolve, reject)
            } catch (err) {
              reject(err)
            }
          })
          _this.onRejectedCallbacks.push(function () {
            try {
              var x = onRejected(_this.value)
              resolutionProcedure(promise2, x, resolve, reject)
            } catch (err) {
              reject(err)
            }
          })
        })
      }

      // 规范 2.3
      /*
      resolutionProcedure函数即为根据x的值来决定promise2的状态的函数
      也即标准中的[Promise Resolution Procedure](https://promisesaplus.com/#point-47)
      x 为 promise2 = promise1.then(onFulfilled, onRejected)里onFulfilled/onRejected的返回值
      resolve 和 reject 实际上是 promise2 的executor的两个实参，因为很难挂在其他地方，所以一并传过来。
      相信各位一定可以对照标准转换成代码，这里就只标出代码在标准中对应的位置，只在必要的地方做一些解释。
      */
      function resolutionProcedure(promise2, x, resolve, reject) {
        // 规范 2.3.1，x 不能和 promise2 相同，避免循环引用
        if (promise2 === x) {
          return reject(new TypeError("Chaining cycle detected for promise!"))
        }
        // 规范 2.3.2
        // 如果 x 为 Promise，状态为 pending 需要继续等待否则执行
        if (x instanceof MyPromise) {
          // 2.3.2.1 如果x为pending状态，promise必须保持pending状态，直到x为fulfilled/rejected
          if (x.currentState === PENDING) {
            x.then(function (value) {
              // 再次调用该函数是为了确认 x resolve 的
              // 参数是什么类型，如果是基本类型就再次 resolve
              // 把值传给下个 then
              resolutionProcedure(promise2, value, resolve, reject)
            }, reject)
          } else { // 但如果这个promise的状态已经确定了，那么它肯定有一个正常的值，而不是一个thenable，所以这里可以取它的状态
            x.then(resolve, reject)
          }
          return
        }

        let called = false
        // 规范 2.3.3，判断 x 是否为对象或函数
        if (x !== null && (typeof x === "object" || typeof x === "function")) {
          // 规范 2.3.3.2，如果不能取出 then，就 reject
          try {
            // 规范2.3.3.1 因为x.then可能是一个getter，这种情况下多次读取就有可能产生副作用
            // 既要判断它的类型，又要调用它，这就是两次读取
            let then = x.then
            // 规范2.3.3.3，如果 then 是函数，调用 x.then
            if (typeof then === "function") {
              // 规范 2.3.3.3
              // reject 或 reject 其中一个执行过的话，忽略其他的
              then.call(
                x,
                y => { // 规范 2.3.3.3.1
                  if (called) return // 规范 2.3.3.3.3，即这三处谁先执行就以谁的结果为准
                  called = true
                  // 规范 2.3.3.3.1
                  return resolutionProcedure(promise2, y, resolve, reject)
                },
                r => {
                  if (called) return // 规范 2.3.3.3.3，即这三处谁先执行就以谁的结果为准
                  called = true
                  return reject(r)
                }
              )
            } else {
              // 规范 2.3.3.4
              resolve(x)
            }
          } catch (e) { // 规范 2.3.3.2
            if (called) return // 规范 2.3.3.3.3，即这三处谁先执行就以谁的结果为准
            called = true
            return reject(e)
          }
        } else {
          // 规范 2.3.4，x 为基本类型
          resolve(x)
        }
      }
    }
    // catch 的实现
    MyPromise.prototype.catch = function (onRejected) {
      return this.then(null, onRejected)
    }
    // finally 的实现
    MyPromise.prototype.finally = function (callback) {
      return this.then(function (value) {
        return MyPromise.resolve(callback()).then(function () {
          return value
        })
      }, function (err) {
        return MyPromise.resolve(callback()).then(function () {
          throw err
        })
      })
    }
    // race
    MyPromise.race = function (values) {
      return new MyPromise(function (resolve, reject) {
        values.forEach(function (value) {
          debugger
          MyPromise.resolve(value).then(resolve, reject)
        })
      })
    }
    // all
    MyPromise.all = function (arr) {
      var args = Array.prototype.slice.call(arr)
      return new MyPromise(function (resolve, reject) {
        if (args.length === 0) return resolve([])
        var remaining = args.length
        for (var i = 0; i < args.length; i++) {
          res(i, args[i])
        }

        function res(i, val) {
          if (val && (typeof val === 'object' || typeof val === 'function')) {
            if (val instanceof MyPromise && val.then === MyPromise.prototype.then) {
              if (val.currentState === FULFILLED) return res(i, val.value)
              if (val.currentState === REJECTED) reject(val.value)
              val.then(function (val) {
                res(i, val)
              }, reject)
              return
            } else {
              var then = val.then
              if (typeof then === 'function') {
                var p = new MyPromise(then.bind(val))
                p.then(function (val) {
                  res(i, val)
                }, reject)
                return
              }
            }
          }
          args[i] = val
          if (--remaining === 0) {
            resolve(args)
          }
        }
      })
    }
    // allSettled
    MyPromise.allSettled = function (promises) {
      return new MyPromise((resolve, reject) => {
        promises = Array.isArray(promises) ? promises : []
        let len = promises.length
        const argslen = len
        // 如果传入的是一个空数组，那么就直接返回一个resolved的空数组promise对象
        if (len === 0) return resolve([])
        // 将传入的参数转化为数组，赋给args变量
        let args = Array.prototype.slice.call(promises)
        // 计算当前是否所有的 promise 执行完成，执行完毕则resolve
        const compute = () => {
          if (--len === 0) {
            resolve(args)
          }
        }

        function resolvePromise(index, value) {
          // 判断传入的是否是 promise 类型
          if (value instanceof MyPromise) {
            const then = value.then
            then.call(value, function (val) {
              args[index] = {
                status: 'fulfilled',
                value: val
              }
              compute()
            }, function (e) {
              args[index] = {
                status: 'rejected',
                reason: e
              }
              compute()
            })
          } else {
            args[index] = {
              status: 'fulfilled',
              value: value
            }
            compute()
          }
        }

        for (let i = 0; i < argslen; i++) {
          resolvePromise(i, args[i])
        }
      })
    }


    const p1 = new MyPromise((resolve, reject) => {
        resolve('hello');
      })
      .then(result => result)
      .catch(e => e);

    const p2 = new MyPromise((resolve, reject) => {
      resolve('jajaj')
        // throw new Error('报错了');
      })
      .then(result => result)
      // .catch(e => e);


    // MyPromise.all([p1, p2])
    //   .then(result => console.log(result))
    //   .catch(e => console.log(e));

    // Promise.all([p1, p2])
    //   .then(result => console.log(result))
    //   .catch(e => console.log(e));

      // MyPromise.race([p1, p2]).then(result => console.log(result)).catch(e => console.log(e))



    // const a = new MyPromise((resolve, reject) => {
    //   console.log('执行')
    //   resolve('aaa')
    // })

  </script> -->
    <script>
      // 判断变量否为function
      const isFunction = (variable) => typeof variable === "function";
      // 定义Promise的三种状态常量
      const PENDING = "PENDING";
      const FULFILLED = "FULFILLED";
      const REJECTED = "REJECTED";

      class MyPromise {
        constructor(handle) {
          if (!isFunction(handle)) {
            throw new Error("MyPromise must accept a function as a parameter");
          }
          // 添加状态
          this._status = PENDING;
          // 添加状态
          this._value = undefined;
          // 添加成功回调函数队列
          this._fulfilledQueues = [];
          // 添加失败回调函数队列
          this._rejectedQueues = [];
          // 执行handle
          try {
            handle(this._resolve.bind(this), this._reject.bind(this));
          } catch (err) {
            this._reject(err);
          }
        }
        // 添加resovle时执行的函数
        _resolve(val) {
          const run = () => {
            if (this._status !== PENDING) return;
            this._status = FULFILLED;
            // 依次执行成功队列中的函数，并清空队列
            const runFulfilled = (value) => {
              let cb;
              while ((cb = this._fulfilledQueues.shift())) {
                cb(value);
              }
            };
            // 依次执行失败队列中的函数，并清空队列
            const runRejected = (error) => {
              let cb;
              while ((cb = this._rejectedQueues.shift())) {
                cb(error);
              }
            };
            /* 如果resolve的参数为Promise对象，则必须等待该Promise对象状态改变后,
            当前Promsie的状态才会改变，且状态取决于参数Promsie对象的状态
          */
            if (val instanceof MyPromise) {
              val.then(
                (value) => {
                  this._value = value;
                  runFulfilled(value);
                },
                (err) => {
                  this._value = err;
                  runRejected(err);
                }
              );
            } else {
              this._value = val;
              runFulfilled(val);
            }
          };
          // 为了支持同步的Promise，这里采用异步调用
          setTimeout(run, 0);
        }
        // 添加reject时执行的函数
        _reject(err) {
          if (this._status !== PENDING) return;
          // 依次执行失败队列中的函数，并清空队列
          const run = () => {
            this._status = REJECTED;
            this._value = err;
            let cb;
            while ((cb = this._rejectedQueues.shift())) {
              cb(err);
            }
          };
          // 为了支持同步的Promise，这里采用异步调用
          setTimeout(run, 0);
        }
        // 添加then方法
        then(onFulfilled, onRejected) {
          const { _value, _status } = this;
          // 返回一个新的Promise对象
          return new MyPromise((onFulfilledNext, onRejectedNext) => {
            // 封装一个成功时执行的函数
            let fulfilled = (value) => {
              try {
                if (!isFunction(onFulfilled)) {
                  onFulfilledNext(value);
                } else {
                  let res = onFulfilled(value);
                  if (res instanceof MyPromise) {
                    // 如果当前回调函数返回MyPromise对象，必须等待其状态改变后在执行下一个回调
                    res.then(onFulfilledNext, onRejectedNext);
                  } else {
                    //否则会将返回结果直接作为参数，传入下一个then的回调函数，并立即执行下一个then的回调函数
                    onFulfilledNext(res);
                  }
                }
              } catch (err) {
                // 如果函数执行出错，新的Promise对象的状态为失败
                onRejectedNext(err);
              }
            };
            // 封装一个失败时执行的函数
            let rejected = (error) => {
              try {
                if (!isFunction(onRejected)) {
                  onRejectedNext(error);
                } else {
                  let res = onRejected(error);
                  if (res instanceof MyPromise) {
                    // 如果当前回调函数返回MyPromise对象，必须等待其状态改变后在执行下一个回调
                    res.then(onFulfilledNext, onRejectedNext);
                  } else {
                    //否则会将返回结果直接作为参数，传入下一个then的回调函数，并立即执行下一个then的回调函数
                    onFulfilledNext(res);
                  }
                }
              } catch (err) {
                // 如果函数执行出错，新的Promise对象的状态为失败
                onRejectedNext(err);
              }
            };
            switch (_status) {
              // 当状态为pending时，将then方法回调函数加入执行队列等待执行
              case PENDING:
                this._fulfilledQueues.push(fulfilled);
                this._rejectedQueues.push(rejected);
                break;
              // 当状态已经改变时，立即执行对应的回调函数
              case FULFILLED:
                fulfilled(_value);
                break;
              case REJECTED:
                rejected(_value);
                break;
            }
          });
        }
        // 添加catch方法
        catch(onRejected) {
          return this.then(undefined, onRejected);
        }
        // 添加静态resolve方法
        static resolve(value) {
          // 如果参数是MyPromise实例，直接返回这个实例
          if (value instanceof MyPromise) return value;
          return new MyPromise((resolve) => resolve(value));
        }
        // 添加静态reject方法
        static reject(value) {
          return new MyPromise((resolve, reject) => reject(value));
        }
        // 添加静态all方法
        static all(list) {
          return new MyPromise((resolve, reject) => {
            /**
             * 返回值的集合
             */
            let values = [];
            let count = 0;
            for (let [i, p] of list.entries()) {
              // 数组参数如果不是MyPromise实例，先调用MyPromise.resolve
              this.resolve(p).then(
                (res) => {
                  values[i] = res;
                  count++;
                  // 所有状态都变成fulfilled时返回的MyPromise状态就变成fulfilled
                  if (count === list.length) resolve(values);
                },
                (err) => {
                  // 有一个被rejected时返回的MyPromise状态就变成rejected
                  reject(err);
                }
              );
            }
          });
        }
        // 添加静态race方法
        static race(list) {
          return new MyPromise((resolve, reject) => {
            for (let p of list) {
              // 只要有一个实例率先改变状态，新的MyPromise的状态就跟着改变
              this.resolve(p).then(
                (res) => {
                  resolve(res);
                },
                (err) => {
                  reject(err);
                }
              );
            }
          });
        }
        finally(cb) {
          return this.then(
            (value) => MyPromise.resolve(cb()).then(() => value),
            (reason) =>
              MyPromise.resolve(cb()).then(() => {
                throw reason;
              })
          );
        }
        // 实现allSettled
        static allSettled(list) {
          return new Promise((resolve, reject) => {
            let values = [];
            let count = 0;
            for (let [i, p] of list.entries()) {
              this.resolve(p)
                .then((res) => {
                  values[count] = res;
                  count++;
                })
                .catch((err) => {
                  values[count] = err;
                  count++;
                })
                .finally(() => {
                  if (count === list.length) resolve(values);
                });
            }
          });
        }

        static all_(list) {
          return new Promise((resove, reject) => {
            const values = [],
              count = 0;
            for (let [i, p] of list.entries()) {
              this.resolve(p).then(
                (res) => {
                  values[i] = res;
                  count++;
                  if (count === list.length) {
                    resove(values);
                  }
                },
                (err) => {
                  reject(err);
                }
              );
            }
          });
        }
      }

      // let myPromise = new MyPromise((resolve, reject) => {
      //   console.log('jajaj')
      //   resolve('resolved')
      // })
      // myPromise.then(res => {
      //   console.log(res)
      // })
    </script>

    <script>
      const someAsyncThing = function () {
        return new Promise(function (resolve, reject) {
          // 下面一行会报错，因为x没有声明
          resolve(x + 2);
        });
      };

      someAsyncThing().then(function () {
        console.log("everything is great");
      });

      setTimeout(() => {
        console.log(123);
      }, 2000);
      // Uncaught (in promise) ReferenceError: x is not defined
      // 123
    </script>
  </body>
</html>
